File: src/239_滑动窗口最大值.rs
// 239_滑动窗口最大值
// https://leetcode.cn/problems/sliding-window-maximum/description/

use std::collections::VecDeque;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let n = nums.len();
        let k = k as usize;
        let mut ans = Vec::new();
        let mut deque = VecDeque::new();

        for i in 0..n {
            // 移除超出窗口范围的元素
            while !deque.is_empty() && deque.front().unwrap() < &(i as i32 - k as i32 + 1) {
                deque.pop_front();
            }

            // 移除比当前元素小的元素，因为它们不可能成为最大值
            while !deque.is_empty() && nums[*deque.back().unwrap() as usize] < nums[i] {
                deque.pop_back();
            }

            deque.push_back(i as i32);

            if i >= k - 1 {
                ans.push(nums[*deque.front().unwrap() as usize]);
            }
        }

        ans
    }
}

File: src/239_滑动窗口最大值.cpp
// 239_滑动窗口最大值
// https://leetcode.cn/problems/sliding-window-maximum/description/

#include <vector>
#include <deque>
#include <iostream>

std::vector<int> maxSlidingWindow(std::vector<int> &nums, int k)
{
    std::vector<int> result;
    std::deque<int> dq; // 存储索引

    for (int i = 0; i < nums.size(); ++i)
    {
        // 移除不在滑动窗口内的元素的索引
        while (!dq.empty() && dq.front() <= i - k)
        {
            dq.pop_front();
        }

        // 维护deque的单调递减性质，移除所有小于当前元素的索引
        while (!dq.empty() && nums[dq.back()] < nums[i])
        {
            dq.pop_back();
        }

        // 将当前元素索引加入deque
        dq.push_back(i);

        // 如果窗口已经达到大小k，记录当前窗口的最大值
        if (i >= k - 1)
        {
            result.push_back(nums[dq.front()]);
        }
    }

    return result;
}

int main()
{
    std::vector<int> nums = {1, 2, 3, 4, 6};
    int k = 2;
    std::vector<int> maxValues = maxSlidingWindow(nums, k);

    for (int val : maxValues)
    {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    return 0;
}